x/tally/keeper/endblock.go

```diff
@@ -199,25 +206,45 @@ func (k Keeper) FilterAndTally(ctx sdk.Context, req types.Request, params types.
 		sort.Strings(reveals[i].ProxyPubKeys)
 	}
 
-	// Phase I: Filtering
+	// Phase 1: Filtering
 	filterResult, filterErr := ExecuteFilter(reveals, req.ConsensusFilter, req.ReplicationFactor, params)
-	result.Consensus = filterResult.Consensus
-	result.ProxyPubKeys = filterResult.ProxyPubKeys
-	result.TallyGasUsed += filterResult.GasUsed
+	// Begin tracking tally result and burn amount.
+	result := TallyResult{
+		Consensus:    filterResult.Consensus,
+		ProxyPubKeys: filterResult.ProxyPubKeys,
+		TallyGasUsed: filterResult.GasUsed,
+	}
+	burnAmt := math.ZeroInt()
+	if filterResult.GasUsed > 0 {
+		burnAmt = gasPrice.Mul(math.NewIntFromUint64(filterResult.GasUsed))
+		ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeTallyGasBurn,
+			sdk.NewAttribute(sdk.AttributeKeyAmount, burnAmt.String()),
+		))
+	}
 
-	// Phase II: Tally Program Execution
+	// Phase 2: Tally Program Execution
+	var vmRes tallyvm.VmResult
+	var tallyErr error
 	if filterErr == nil {
-		vmRes, err := k.ExecuteTallyProgram(ctx, req, filterResult, reveals)
-		if err != nil {
-			result.Result = []byte(err.Error())
+		vmRes, tallyErr = k.ExecuteTallyProgram(ctx, req, filterResult, reveals)
```

Comment:
The execution needs to be aware of the base gas fee that was already consumed.

Reply:
Shouldn't the base gas be taken from the execution gas limit while the tally program uses the tally gas limit?

Reply:
OK added. Perhaps `BuildFilter()` should check if the filter gas cost is within the tally limit?

Reply:
I sort of addressed this in the wall of text about gas metering. TL;DR: we're missing a few gas checks before we start an operation.